File: coding/src/com/algos/topics/li11_heap/revised/L973_KClosestPointsToOrigin.java
package  com.algos.topics.li11_heap.revised;

import java.util.Arrays;
import java.util.PriorityQueue;
import java.util.Queue;

public class L973_KClosestPointsToOrigin {

	public static void main(String[] args) {
		L973_KClosestPointsToOrigin obj = new L973_KClosestPointsToOrigin();
		int[][] points = new int[][] {{6,10},{-3,3},{-2,5},{0,2}};
		int[][] result = obj.kClosest(points, 3);
		for(int[] point: result)
			System.out.println(Arrays.toString(point));
	}

	// Problem: https://leetcode.com/problems/k-closest-points-to-origin/
	// Idea: Load all the points with sorter as eulidian distance then pop k points from min heap.
    public int[][] kClosest(int[][] points, int k) {
		Queue<int[]> maxHeap =  new PriorityQueue<int[]>((p1, p2) -> {
			return (p2[0]*p2[0]+p2[1]*p2[1]) - (p1[0]*p1[0]+p1[1]*p1[1]);
        });

        for(int[] point: points) {
			maxHeap.add(point);
			if (maxHeap.size() > k) {
                maxHeap.remove();
            }
		}

		int[][] result = new int[k][2];
        while (k > 0) {
            result[--k] = maxHeap.poll();
        }

        return result;
    }
}

File: dsa/patterns/11_heaps/python/l973_kclosest_points_to_origin.py
import heapq

"""
Problem: https://leetcode.com/problems/k-closest-points-to-origin/
Idea: Maintain a max heap of size k based on squared Euclidean distance.
      If heap size exceeds k, remove the farthest point.
"""
def k_closest(points, k):
    max_heap = []

    for x, y in points:
        dist = x*x + y*y
        heapq.heappush(max_heap, (-dist, x, y))
        if len(max_heap) > k:
            heapq.heappop(max_heap)

    result = []
    while max_heap:
        _, x, y = heapq.heappop(max_heap)
        result.append([x, y])

    return result


# ---- one test (similar to Java main) ----
if __name__ == "__main__":
    points = [[6, 10], [-3, 3], [-2, 5], [0, 2]]
    res = k_closest(points, 3)
    for p in res:
        print(p)
